### PR TITLE
Repanova rebased

### PR DESCRIPTION
**Pull request by josef-pkt**
this is PR #3303 after interactive rebase to remove commits from master

needs another check, no merge conflicts in rebase, but I had to remove the majority of commits.

### CODE REVIEW

**Comment by josef-pkt** [comment]
In direct file comparison the two files in here are identical to the ones in 3303, so rebase is good

File: statsmodels/stats/anova.py [hunk 1]
```diff
@@ -362,6 +367,197 @@ def anova_lm(*args, **kwargs):
...
+    for term in slices_to_exclude:
+        s = slices[term]
+        ind[s] = False
+    return ind
+
+
+def _ssr_reduced_model(y, x, term_slices, params, keys):
+    """
+    Residual sum of squares of OLS model excluding factors in `keys`
+    Assumes x matrix is orthogonal
```
Comment (josef-pkt) [inline statsmodels/stats/anova.py]: blank line after summary line

File: statsmodels/stats/anova.py [hunk 2]
```diff
@@ -362,6 +367,197 @@ def anova_lm(*args, **kwargs):
...
+
+    Parameters
+    ----------
+    y : array_like, dependent variables
+    x : array_like, independent variables
```
Comment (josef-pkt) [inline statsmodels/stats/anova.py]: use two lines, `independent variables` in second line, doc standard

File: statsmodels/stats/anova.py [hunk 3]
```diff
@@ -362,6 +367,197 @@ def anova_lm(*args, **kwargs):
...
+    term_slices : a dict of slices
+        term_slices[key] is a boolean array specifies the parameters
+        associated with the factor `key`
+    params : OLS solution of y = x * params
+    keys : factors to be excluded
+
+    Returns
+    -------
+    residual sum of squares and degrees of freedom
```
Comment (josef-pkt) [inline statsmodels/stats/anova.py]: list for returns

File: statsmodels/stats/anova.py [hunk 4]
```diff
@@ -362,6 +367,197 @@ def anova_lm(*args, **kwargs):
...
+    .. [1] Rutherford, Andrew. Anova and ANCOVA: a GLM approach. John Wiley & Sons, 2011.
+
+    """
+
+    def __init__(self, data, depvar, subject, within=None, between=None):
+        self.data = data
+        self.depvar = depvar
+        self.within = within
+        if 'C' in within:
+            raise ValueError("Factor name cannot be 'C'! This is in conflict "
```
Comment (josef-pkt) [inline statsmodels/stats/anova.py]: is this check too strict?
is `"Cat"` allowed?

File: statsmodels/stats/anova.py [hunk 5]
```diff
@@ -362,6 +367,197 @@ def anova_lm(*args, **kwargs):
...
+                             "with patsy's contrast function name.")
+        self.between = between
+        if between is not None:
+            raise NotImplementedError('Between subject effect not '
+                                      'yet supported!')
+        self.subject = subject
+        self._check_data_balanced()
+
+    def _check_data_balanced(self):
```
Comment (josef-pkt) [inline statsmodels/stats/anova.py]: docstring missing
I don't see a return. If no exception is raised, then data is balanced?

File: statsmodels/stats/anova.py [hunk 6]
```diff
@@ -362,6 +367,197 @@ def anova_lm(*args, **kwargs):
...
+            raise ValueError(error_message)
+        count = cell_count[key]
+        for key in cell_count:
+            if count != cell_count[key]:
+                raise ValueError(error_message)
+        if self.data.shape[0] > count * factor_levels:
+            raise ValueError('There are more than 1 element in a cell! Missing'
+                             ' factors?')
+
+    def fit(self):
```
Comment (josef-pkt) [inline statsmodels/stats/anova.py]: docstring so it gets added to the docs even if it doesn't have any arguments

File: statsmodels/stats/tests/test_anova_rm.py [hunk 7]
```diff
@@ -0,0 +1,133 @@
...
+    a = [[1, 7, 22.4, 0.002125452]]
+    assert_array_almost_equal(df.anova_table.iloc[:, [1, 2, 0, 3]].values,
+                              a, decimal=5)
+
+
+def test_two_factors_repeated_measures_anova():
+    """
+    Testing two factors repeated measures anova
+    Results reproduces R `ezANOVA` function from library ez
+    """
```
Comment (josef-pkt) [inline statsmodels/stats/tests/test_anova_rm.py]: no docstrings for unit test functions, use code comment `#`

File: statsmodels/stats/anova.py [hunk 8]
```diff
@@ -362,6 +367,197 @@ def anova_lm(*args, **kwargs):
...
+    ----------
+    anova_table : DataFrame
+    """
+    def __init__(self, anova_table):
+        self.anova_table = anova_table
+
+    def __str__(self):
+        return self.summary().__str__()
+
+    def summary(self, contrast_L=False, transform_M=False):
```
Comment (josef-pkt) [inline statsmodels/stats/anova.py]: contrast_L and transform_M are not used in the function

**Comment by josef-pkt** [comment]
While trying to write docstrings: Some arguments or returns of mainly the internal functions are difficult to understand. e.g. What's the structure term_slices and what type are the keys?
I have no idea from reading the code, without having played with examples yet.
(a guess would be that the keys are strings with factor names)

```
    term_slices : a dict of slices
        term_slices[key] is a boolean array specifies the parameters
        associated with the factor `key`
    ...
    keys : keys for term_slices
        factors to be excluded
```

**Comment by josef-pkt** [comment]
I had rebased on outdated local master instead of current statsmodels master.
rebased again and force pushed

**Comment by josef-pkt** [comment]
I would like to merge this as is, after test run comes back green
If there is feedback until the release, then we can still make changes.

@yl565 @jona-sassenhagen Thanks for code and review, and sorry about my slow work.
Please try it out and provide feedback, especially if you see things or API that can be improved.

**Comment by josef-pkt** [comment]
all green (appveyor has one with the usual statespace problem)

merging

**Comment by jona-sassenhagen** [comment]
Thanks @josef-pkt !

**Comment by yl565** [comment]
Thanks!

**Pull request merged**